File: Leetcode_Code/235. Lowest Common Ancestor of a Binary Search Tree/235.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (!root) return NULL;
        int mx = max(p->val, q->val);
        int mn = min(p->val, q->val);
        while (true) {
            int value = root->val;
            if (value <= mx && value >= mn) return root;
            if (value < mn) root = root->right;
            if (value > mx) root = root->left;
        }
    }
};

File: Leetcode_Code/235. Lowest Common Ancestor of a Binary Search Tree/235.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if root is None:
            return None
        mx = max(p.val, q.val)
        mn = min(p.val, q.val)
        while True:
            value = root.val
            if value <= mx and value >= mn:
                return root
            if value < mn:
                root = root.right
            if value > mx:
                root = root.left
